libpromises/eval_context.c

```diff
@@ -3847,3 +3902,180 @@ bool EvalContextOverrideImmutableGet(EvalContext *ctx)
+        SeqAppend(parent_event_frame->event->children, last_event);
+        last_event->parent = parent_event_frame->event;
+    }
+    else
+    {
+        SeqAppend(ctx->root->children, last_event);
+        last_event->parent = ctx->root;
+    }
+}
+
+static EventFrame *BundleToEventFrame(const Bundle *bp)
+{
+    assert(bp != NULL);
+    return EventFrameCreate("bundle", bp->name, bp->source_path, bp->offset);
+}
+
+static EventFrame *PromiseToEventFrame(const Promise *pp)
+{
+    assert(pp != NULL);
+    return EventFrameCreate("promise", PromiseGetPromiseType(pp), PromiseGetBundle(pp)->source_path, pp->offset);
+}
+
+static EventFrame *FunctionToEventFrame(const FnCall *fp)
+{
+    assert(fp != NULL);
+    return EventFrameCreate("function", fp->name, PromiseGetBundle(fp->caller)->source_path, fp->caller->offset);
+}
+
+static EventFrame *PolicyToEventFrame()
+{
+    return EventFrameCreate("policy", "", "", (SourceOffset) {0});
+}
+
+void EvalContextAddFunctionEvent(EvalContext *ctx, const FnCall *fp, int64_t start)
+{
+    assert(ctx != NULL);
+    if (ctx->profiling)
+    {
+        EventFrame *event = FunctionToEventFrame(fp);
+        int64_t end = EvalContextEventStart();
+        assert(end >= start); /* sanity check */
+
+        event->elapsed = end - start;
+        EventFrameUpdate(ctx, event);
+    }
+}
+
+static JsonElement *EventFrameTreeToJson(const EventFrame *node)
+{
+    assert(node != NULL);
+    JsonElement *new_element = JsonObjectCreate(6);
+
+    JsonObjectAppendString(new_element, "name", node->name);
+    JsonObjectAppendString(new_element, "type", node->type);
+    JsonObjectAppendString(new_element, "filename", node->filename);
+    JsonObjectAppendInteger64(new_element, "elapsed", node->elapsed);
+    
+    JsonElement *offset = JsonObjectCreate(3);
+    JsonObjectAppendInteger(offset, "lineno", node->offset.line);
+    JsonObjectAppendInteger(offset, "start", node->offset.start);
+    JsonObjectAppendInteger(offset, "end", node->offset.end);
+    JsonObjectAppendObject(new_element, "offset", offset);
+
+    JsonElement *children = JsonArrayCreate(5);
+
+    const size_t length = SeqLength(node->children);
+    for (size_t i = 0; i < length; i++)
+    {
+        EventFrame *next = SeqAt(node->children, i);
+        JsonElement *child = EventFrameTreeToJson(next);
+        if (child != NULL)
+        {
+            JsonArrayAppendObject(children, child);
+        }
+    }
+    JsonObjectAppendArray(new_element, "children", children);
+    return new_element;
+}
+
+int64_t EvalContextEventStart()
+{
+    struct timespec ts;
+    clock_gettime(CLOCK_MONOTONIC, &ts);
+    return ts.tv_sec * 1000000000LL + ts.tv_nsec;
+}
+
+void EvalContextWriteEventTree(EvalContext *ctx, int64_t start)
+{
+    assert(ctx != NULL);
+    int64_t end = EvalContextEventStart();
+    ctx->root->elapsed = end - start;
+
+    if (!ctx->profiling)
+    {
+        return;
+    }
+
+    JsonElement *tree = EventFrameTreeToJson(ctx->root);
```

Comment:
Here we should loop over the "unprocessed" events in the Seq, and convert them to the proper tree.